psutil/_psibmi.py

```diff
@@ -0,0 +1,647 @@
+        # process is no longer active so we force NSP in case the PID
+        # is no longer there.
+        # TODO IBM i implementation
+        # if not ret:
+        #     # will raise NSP if process is gone
+        #     os.stat('%s/%s' % (self._procfs_path, self.pid))
+        return ret
+
+    @wrap_exceptions
+    def nice_get(self):
+        return cext_posix.getpriority(self.pid)
+
+    @wrap_exceptions
+    def nice_set(self, value):
+        return cext_posix.setpriority(self.pid, value)
+
+    @wrap_exceptions
+    def ppid(self):
+        self._ppid = self._proc_basic_info()[proc_info_map['ppid']]
+        return self._ppid
+
+    @wrap_exceptions
+    def uids(self):
+        real, effective, saved, _, _, _ = self._proc_cred()
+        return _common.puids(real, effective, saved)
+
+    @wrap_exceptions
+    def gids(self):
+        _, _, _, real, effective, saved = self._proc_cred()
+        return _common.puids(real, effective, saved)
+
+    @wrap_exceptions
+    def cpu_times(self):
+        cpu_times = cext.proc_cpu_times(self.pid)
+        return _common.pcputimes(*cpu_times)
+
+    @wrap_exceptions
+    def terminal(self):
+
+        p = subprocess.Popen(
+            ["/QOpenSys/usr/bin/ps",
+             "-Af",
+             "-o",
+             "tty,pid,user",
+             "-p", str(self.pid)], stdout=subprocess.PIPE)
+        output = p.communicate()[0]
+        for line in output.decode("utf-8") .splitlines()[1:]:
+            tokens = line.split()
+            tty = tokens[0]
+            pid = int(tokens[1])
+            if(tty == "-"):
+                continue
+            if(pid == int(self.pid)):
+                return "/dev/"+tty
+        return None
+
+    @wrap_exceptions
+    def cwd(self):
+        _not_supported()  # TODO
+
+    @wrap_exceptions
+    def memory_info(self):
+        ret = self._proc_basic_info()
+        rss = ret[proc_info_map['rss']] * 1024
+        vms = ret[proc_info_map['vms']] * 1024
+        return pmem(rss, vms)
+
+    memory_full_info = memory_info
+
+    @wrap_exceptions
+    def status(self):
+        code = self._proc_basic_info()[proc_info_map['status']]
+        # XXX is '?' legit? (we're not supposed to return it anyway)
+        return PROC_STATUSES.get(code, '?')
+
+    def open_files(self):
+        # TODO rewrite without using procfiles (stat /proc/pid/fd/* and then
+        # find matching name of the inode)
+        p = subprocess.Popen(["/usr/bin/procfiles", "-n", str(self.pid)],
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = p.communicate()
+        if PY3:
+            stdout, stderr = [x.decode(sys.stdout.encoding)
+                              for x in (stdout, stderr)]
+        if "no such process" in stderr.lower():
+            raise NoSuchProcess(self.pid, self._name)
+        procfiles = re.findall(r"(\d+): S_IFREG.*\s*.*name:(.*)\n", stdout)
```

Comment:
You can probably use `re.compile('...')` and use it as a private argument of this function (faster). 
Grep for "re.compile" to see what I mean.